File: sdks/jvm/src/main/kotlin/com/heossi/qnsi/KmsClient.kt
package com.heossi.qnsi

import com.heossi.qnsi.internal.Internal
import com.heossi.qnsi.internal.RequestOptions
import com.heossi.qnsi.internal.decodeB64
import com.heossi.qnsi.internal.encodeB64
import com.heossi.qnsi.internal.boolField
import com.heossi.qnsi.internal.stringField
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

private const val PREFIX = "/proxy/kms/v1"

/** Create a server-side PQC key. `purpose` is one of `"signing"`, `"encryption"`, `"kem"`. */
@Serializable
public data class CreateKeyRequest @JvmOverloads constructor(
    public val algorithm: String,
    public val purpose: String,
    public val metadata: JsonObject? = null,
)

/** Server-side PQC keys: create / rotate / sign / verify / wrap / unwrap. Wraps kms-service `/kms/v1`. */
public class KmsClient internal constructor(private val transport: Internal) {
    @JvmOverloads
    public fun createKey(req: CreateKeyRequest, idempotencyKey: String? = null): JsonObject {
        // Backend createKeySchema = { tenantId(auto), keyId(required), keyType(required:
        // root|master|data|byok), algorithm, metadata }. The SDK's `purpose` is NOT a backend
        // field, so fold it into metadata, generate a keyId, and default keyType to "data"
        // (createKey previously serialized {algorithm, purpose} with no keyId/keyType -> 400).
        val meta = buildJsonObject {
            req.metadata?.forEach { (k, v) -> put(k, v) }
            put("purpose", req.purpose)
        }
        val body = buildJsonObject {
            put("keyId", java.util.UUID.randomUUID().toString())
            put("keyType", "data")
            put("algorithm", req.algorithm)
            put("metadata", meta)
        }.toString()
        return transport.requestJson(
            "POST", "$PREFIX/keys", body,
            RequestOptions(idempotencyKey = idempotencyKey),
        )
    }

    @JvmOverloads
    public fun listKeys(query: Map<String, String?> = emptyMap()): JsonObject =
        transport.requestJson("GET", "$PREFIX/keys", null, RequestOptions(query = query))

    public fun getKey(keyId: String): JsonObject =
        transport.requestJson("GET", "$PREFIX/keys/$keyId", null)

    @JvmOverloads
    public fun rotateKey(keyId: String, idempotencyKey: String? = null): JsonObject =
        transport.requestJson(
            "POST", "$PREFIX/keys/$keyId/rotate", null,
            RequestOptions(idempotencyKey = idempotencyKey),
        )

    public fun deleteKey(keyId: String): JsonObject =
        transport.requestJson("DELETE", "$PREFIX/keys/$keyId", null)

    @JvmOverloads
    public fun sign(keyId: String, data: ByteArray, idempotencyKey: String? = null): ByteArray {
        val body = buildJsonObject { put("data", encodeB64(data)) }.toString()
        val resp = transport.requestJson(
            "POST", "$PREFIX/keys/$keyId/sign", body,
            RequestOptions(idempotencyKey = idempotencyKey),
        )
        return decodeField(resp, "signature", "sign")
    }

    public fun verify(keyId: String, data: ByteArray, signature: ByteArray): Boolean {
        val body = buildJsonObject {
            put("data", encodeB64(data))
            put("signature", encodeB64(signature))
        }.toString()
        val resp = transport.requestJson("POST", "$PREFIX/keys/$keyId/verify", body)
        return resp.boolField("valid") == true
    }

    @JvmOverloads
    public fun wrap(keyId: String, plaintext: ByteArray, idempotencyKey: String? = null): ByteArray {
        val body = buildJsonObject { put("dataKey", encodeB64(plaintext)) }.toString()
        val resp = transport.requestJson(
            "POST", "$PREFIX/keys/$keyId/wrap", body,
            RequestOptions(idempotencyKey = idempotencyKey),
        )
        return decodeField(resp, "wrappedKey", "wrap")
    }

    @JvmOverloads
    public fun unwrap(keyId: String, ciphertext: ByteArray, idempotencyKey: String? = null): ByteArray {
        val body = buildJsonObject { put("wrappedKey", encodeB64(ciphertext)) }.toString()
        val resp = transport.requestJson(
            "POST", "$PREFIX/keys/$keyId/unwrap", body,
            RequestOptions(idempotencyKey = idempotencyKey),
        )
        return decodeField(resp, "dataKey", "unwrap")
    }

    private fun decodeField(resp: JsonObject, field: String, op: String): ByteArray {
        val b64 = resp.stringField(field)
            ?: throw QnsiApiException("kms.$op: response missing $field", 200)
        return decodeB64(b64)
            ?: throw QnsiApiException("kms.$op: $field is not valid base64", 200)
    }
}

File: sdks/swift/Sources/QNSI/Clients/KmsClient.swift
import Foundation

private let prefix = "/proxy/kms/v1"

/// Server-side PQC keys: create / rotate / sign / verify / wrap / unwrap.
/// Wraps kms-service `/kms/v1`.
public struct KmsClient: Sendable {
    let transport: Transport

    /// Create a server-side PQC key. `purpose` is one of `"signing"`, `"encryption"`, `"kem"`.
    /// The backend `createKeySchema` is `{ tenantId(auto), keyId(required), keyType(required),
    /// algorithm, metadata }`; `purpose` is folded into metadata, a keyId is generated,
    /// and keyType defaults to `"data"` (mirrors the JVM SDK's verified wire contract).
    public func createKey(
        algorithm: String,
        purpose: String,
        metadata: [String: JSONValue]? = nil,
        idempotencyKey: String? = nil
    ) async throws -> JSONValue {
        var meta = metadata ?? [:]
        meta["purpose"] = .string(purpose)
        let body: JSONValue = .object([
            "keyId": .string(UUID().uuidString.lowercased()),
            "keyType": .string("data"),
            "algorithm": .string(algorithm),
            "metadata": .object(meta),
        ])
        return try await transport.requestJson(
            "POST", "\(prefix)/keys", body: body,
            options: RequestOptions(idempotencyKey: idempotencyKey)
        )
    }

    public func listKeys(query: [String: String?] = [:]) async throws -> JSONValue {
        try await transport.requestJson("GET", "\(prefix)/keys", options: RequestOptions(query: query))
    }

    public func getKey(keyId: String) async throws -> JSONValue {
        try await transport.requestJson("GET", "\(prefix)/keys/\(keyId)")
    }

    public func rotateKey(keyId: String, idempotencyKey: String? = nil) async throws -> JSONValue {
        try await transport.requestJson(
            "POST", "\(prefix)/keys/\(keyId)/rotate",
            options: RequestOptions(idempotencyKey: idempotencyKey)
        )
    }

    public func deleteKey(keyId: String) async throws -> JSONValue {
        try await transport.requestJson("DELETE", "\(prefix)/keys/\(keyId)")
    }

    /// Sign `data` with a server-side key. Request field is `data` (base64); the
    /// response field is `signature` (base64).
    public func sign(keyId: String, data: Data, idempotencyKey: String? = nil) async throws -> Data {
        let body: JSONValue = .object(["data": .string(encodeB64(data))])
        let resp = try await transport.requestJson(
            "POST", "\(prefix)/keys/\(keyId)/sign", body: body,
            options: RequestOptions(idempotencyKey: idempotencyKey)
        )
        return try decodeField(resp, field: "signature", op: "sign")
    }

    /// Verify a signature with a server-side key. Returns the backend's `valid` verdict.
    public func verify(keyId: String, data: Data, signature: Data) async throws -> Bool {
        let body: JSONValue = .object([
            "data": .string(encodeB64(data)),
            "signature": .string(encodeB64(signature)),
        ])
        let resp = try await transport.requestJson("POST", "\(prefix)/keys/\(keyId)/verify", body: body)
        return resp["valid"]?.boolValue == true
    }

    /// Wrap a data key. Request field is `dataKey`; response field is `wrappedKey`.
    public func wrap(keyId: String, plaintext: Data, idempotencyKey: String? = nil) async throws -> Data {
        let body: JSONValue = .object(["dataKey": .string(encodeB64(plaintext))])
        let resp = try await transport.requestJson(
            "POST", "\(prefix)/keys/\(keyId)/wrap", body: body,
            options: RequestOptions(idempotencyKey: idempotencyKey)
        )
        return try decodeField(resp, field: "wrappedKey", op: "wrap")
    }

    /// Unwrap a wrapped data key. Request field is `wrappedKey`; response field is `dataKey`.
    public func unwrap(keyId: String, ciphertext: Data, idempotencyKey: String? = nil) async throws -> Data {
        let body: JSONValue = .object(["wrappedKey": .string(encodeB64(ciphertext))])
        let resp = try await transport.requestJson(
            "POST", "\(prefix)/keys/\(keyId)/unwrap", body: body,
            options: RequestOptions(idempotencyKey: idempotencyKey)
        )
        return try decodeField(resp, field: "dataKey", op: "unwrap")
    }

    private func decodeField(_ resp: JSONValue, field: String, op: String) throws -> Data {
        guard let b64 = resp[field]?.stringValue else {
            throw QnsiError.api(message: "kms.\(op): response missing \(field)", statusCode: 200, code: nil, body: nil)
        }
        guard let bytes = decodeB64(b64) else {
            throw QnsiError.api(message: "kms.\(op): \(field) is not valid base64", statusCode: 200, code: nil, body: nil)
        }
        return bytes
    }
}
